acorn/src/regexp.js

```diff
@@ -920,6 +981,249 @@ pp.regexp_eatClassEscape = function(state) {
   )
 }
 
+// https://tc39.es/ecma262/#prod-ClassSetExpression
+// https://tc39.es/ecma262/#prod-ClassUnion
+// https://tc39.es/ecma262/#prod-ClassIntersection
+// https://tc39.es/ecma262/#prod-ClassSubtraction
+pp.regexp_classSetExpression = function(state) {
+  let nextMayContainStrings = false
+  if (this.regexp_eatClassSetRange(state))
+    ;
+  else if (this.regexp_eatClassSetOperand(state)) {
+    nextMayContainStrings = state.lastMayContainStrings
+    let pos = state.pos
+    const start = pos
+    // https://tc39.es/ecma262/#prod-ClassIntersection
+    while (
+      state.eat(0x26 /* & */) && state.eat(0x26 /* & */) &&
+      state.current() !== 0x26 /* & */ &&
+      this.regexp_eatClassSetOperand(state)
+    ) {
+      if (!state.lastMayContainStrings) nextMayContainStrings = false
+      pos = state.pos
+    }
+    state.pos = pos
+    if (state.pos !== start) {
+      state.lastMayContainStrings = nextMayContainStrings
+      return
+    }
+    // https://tc39.es/ecma262/#prod-ClassSubtraction
+    while (
+      state.eat(0x2D /* - */) && state.eat(0x2D /* - */) &&
+      this.regexp_eatClassSetOperand(state)
+    ) {
+      pos = state.pos
+    }
+    state.pos = pos
+    if (state.pos !== start) {
+      state.lastMayContainStrings = nextMayContainStrings
+      return
+    }
+  } else {
+    const ch = state.current()
+    if (ch === 0x5C /* \ */ && state.lookahead() !== 0x71 /* q */) {
+      // Make the same message as V8.
+      state.raise("Invalid escape")
+    }
+    if (ch === state.lookahead() && isClassSetReservedDoublePunctuatorCharacter(ch)) {
+      // Make the same message as V8.
+      state.raise("Invalid set operation in character class")
+    }
+    state.raise("Invalid character in character class")
+  }
+  // https://tc39.es/ecma262/#prod-ClassUnion
+  for (;;) {
+    if (this.regexp_eatClassSetRange(state)) {
+      continue
+    }
+    if (this.regexp_eatClassSetOperand(state)) {
+      if (state.lastMayContainStrings) nextMayContainStrings = true
+      continue
+    }
+    break
+  }
+  state.lastMayContainStrings = nextMayContainStrings
+}
+
+// https://tc39.es/ecma262/#prod-ClassSetRange
+pp.regexp_eatClassSetRange = function(state) {
+  const start = state.pos
+  if (this.regexp_eatClassSetCharacter(state)) {
+    const left = state.lastIntValue
+    if (state.eat(0x2D /* - */) && this.regexp_eatClassSetCharacter(state)) {
```

Comment:
Here too the side effect of `eat` seems to 'leak' when the condition is false.

Reply:
My bad—I forgot to delete this comment after spotting the `pos` reset. Still not a fan of this pattern though, see my other comment.

Reply:
I think I understand that you don't like `state.pos = start`. But for this part, I don't think it makes sense to do it any other way.
The way I think without `eat` is to do:

```js
pp.regexp_eatClassSetRange = function(state) {
  const start = state.pos
  if (this.regexp_eatClassSetCharacter(state)) {
    const left = state.lastIntValue
    if (state.current() === 0x2D /* - */) {
      state.advance()
      if (this.regexp_eatClassSetCharacter(state)) {
        const right = state.lastIntValue
        if (left !== -1 && right !== -1 && left > right) {
          state.raise("Range out of order in character class")
        }
        return true
      }
    }
    state.pos = start
  }
  return false
}
```

`state.current()`+`state.advance()` is consequently the same as `eat()`, and `state.pos = start` cannot be removed yet. If you know a better way, could you please advise?